yeat/tests/test_config.py

```diff
@@ -7,44 +7,88 @@
 # Development Center.
 # -------------------------------------------------------------------------------------------------
 
+import json
+from pathlib import Path
 import pytest
-from yeat.cli.config import AssemblerConfig, AssemblyConfigurationError
+from yeat.cli.config import (
+    Assembler,
+    AssemblerConfig,
+    AssemblyConfigurationError,
+    Sample,
+    KEYS1,
+    KEYS2,
+)
 from yeat.tests import data_file
 
 
 def test_unsupported_assembly_algorithm():
     algorithm = "unsupported_algorithm"
     pattern = rf"Unsupported assembly algorithm '{algorithm}'"
     with pytest.raises(ValueError, match=pattern):
-        AssemblerConfig(algorithm)
+        Assembler("label1", algorithm, ["sample1"])
 
 
-def test_duplicate_assembly_algorithms():
-    pattern = r"Duplicate assembly configuration: please check config file"
+def test_duplicate_assembly_labels(tmp_path):
+    wd = str(tmp_path)
+    data = json.load(open(data_file("configs/paired.cfg")))
+    data["assemblers"].append(data["assemblers"][0])
+    print(data["assemblers"])
+    cfg = str(Path(wd).resolve() / "paired.cfg")
+    json.dump(data, open(cfg, "w"))
+    pattern = r"Duplicate assembly labels: please check config file"
     with pytest.raises(ValueError, match=pattern):
-        AssemblerConfig.parse_json(open(data_file("dup_algorithms.cfg")))
+        AssemblerConfig.parse_json(open(cfg))
 
 
 def test_parse_json():
-    f = open(data_file("config.cfg"))
-    assemblers = [x.algorithm for x in AssemblerConfig.parse_json(f)]
-    assert assemblers == ["spades", "megahit"]
+    f = open(data_file("configs/paired.cfg"))
+    samples, algorithms = AssemblerConfig.parse_json(f)
+    for sample_name, reads in samples.items():
+        assert len(reads.sample) == 2
+    assemblers = [assembler.algorithm for assembler in algorithms]
+    assert assemblers == ["spades", "megahit", "unicycler"]
 
 
 def test_valid_config_entry():
-    data = {"algorithm": "spades", "extra_args": ""}
-    AssemblerConfig.validate(data)
+    data = json.load(open(data_file("configs/paired.cfg")))
+    AssemblerConfig.validate(data, KEYS1)
+    for assembler in data["assemblers"]:
+        AssemblerConfig.validate(assembler, KEYS2)
```

Comment:
The awkwardness of this code is what I was referring to earlier. Wouldn't something like this be clearer?

```python
config = AssemblerConfig.from_json(infile)
config.validate()
```

Reply:
The `validate()` is now called in the `AssemblerConfig::__init__()`.